kdb/keyset.go

```diff
@@ -135,37 +151,73 @@ func (ks *ckeySet) Cut(key Key) KeySet {
 		return nil
 	}
 
-	newKs := C.ksCut(ks.ptr, k.ptr)
+	newKs := wrapKeySet(C.ksCut(ks.ptr, k.ptr))
+
+	newKs.loop(func(key Key) {
+		ks.forgetKey(k.ptr)
+	})
 
-	return wrapKeySet(newKs)
+	return newKs
 }
 
-// Slice returns a slice containing Keys.
+// Slice returns a slice containing all Keys.
 func (ks *ckeySet) Slice() []Key {
 	var keys []Key
 
-	ks.loop(func(k Key) bool {
+	ks.loop(func(k Key) {
 		keys = append(keys, k)
-		return true
 	})
 
 	return keys
 }
 
 // Iterator is a function that loops over Keys.
-// Return false if you want break out of the loop.
-type Iterator func(k Key) bool
+type Iterator func(k Key)
 
-func (ks *ckeySet) loop(iterator Iterator) {
-	for cursor := C.cursor_t(0); C.ksAtCursor(ks.ptr, cursor) != nil; cursor++ {
-		key := wrapKey(C.ksAtCursor(ks.ptr, cursor))
-		cont := iterator(key)
+// toKey returns a cached Key that wraps the *C.struct__Key -
+// or creates a new wrapped *ckey.
+func (ks *ckeySet) toKey(k *C.struct__Key) *ckey {
+	if k == nil {
+		return nil
+	}
 
-		if !cont {
-			break
-		}
+	if key := ks.keys[k]; key != nil {
+		return key
+	} else {
+		return wrapKey(k)
 	}
+}
 
+// rememberKey remembers the relationship between instances of *ckey
```

Comment:
Document more the user view and less the implementation details.

Reply:
There is no user-view, as this is an unexported method. I think it's important for future developers of this package to know why this is done since it's not self-explanatory.

Reply:
Ahh, ok. How do I see which symbols are exported?

Reply:
Capitalized functions are exported, lower-case are not